File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode currP = root;
        TreeNode currQ = root;
        TreeNode prev = null;
        while (currP != null && currQ != null) {
            prev = currP;
            if (currP.val == p.val) {
                return currP;
            }
            if (currQ.val == q.val) {
                return currQ;
            }
            currP = search(currP, p);
            currQ = search(currQ, q);
            if (currP != currQ) {
                return prev;
            }
        }
        return null;
    }

    private TreeNode search(TreeNode node, TreeNode p) {
        if (node == null) {
            return null;
        }

        if (node.val == p.val) {
            return node;
        } else if (node.val > p.val) {
            return node.left;
        } else {
            return node.right;
        }
    }
}

File: Data Structures & Algorithms/house-robber-iii/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode]) -> List[int]:
            if not node:
                return [0, 0]
            left = dfs(node.left)
            right = dfs(node.right)
            node_robbed = node.val + left[1] + right[1]
            node_not_robbed = max(left) + max(right)
            return [node_robbed, node_not_robbed]
        return max(dfs(root)) 